File: src/ts/input-validator.ts
/**
 * Input validation and sanitization module for BC Calculator
 * Prevents command injection and ensures safe input to BC
 */

import { ValidationResult } from './types.js';

export class InputValidator {
  // Maximum expression length (10KB)
  private static readonly MAX_LENGTH = 10000;
  
  // Allowed characters: numbers, letters, math operators, BC syntax
  private static readonly ALLOWED_CHARS = 
    /^[0-9a-zA-Z+\-*\/^().,;\s=<>!&|%{}[\]]+$/;
  
  // Dangerous patterns that could lead to command injection or system abuse
  private static readonly DANGEROUS_PATTERNS = [
    /system\s*\(/i,           // system() calls
    /exec\s*\(/i,             // exec() calls
    /`/,                      // Backticks (command substitution)
    /\$\(/,                   // Command substitution
    />\s*[\/\w]/,             // File redirects (output)
    /<\s*[\/\w]/,             // File redirects (input)
    /\|\s*[\/\w]/,            // Pipes to other commands
    /;\s*(?:bash|sh|rm|cat|ls|pwd|chmod|chown)/i, // Dangerous shell commands
  ];

  /**
   * Validate an expression for safety and correctness
   * @param expression The expression to validate
   * @returns ValidationResult with validation status and optional error message
   */
  static validate(expression: string): ValidationResult {
    // 1. Check for null/undefined
    if (!expression) {
      return {
        valid: false,
        error: 'Expression cannot be null or undefined'
      };
    }

    // 2. Check length
    if (expression.length > this.MAX_LENGTH) {
      return {
        valid: false,
        error: `Expression too long (max ${this.MAX_LENGTH} characters, got ${expression.length})`
      };
    }

    // 3. Check for empty input after trimming
    const trimmed = expression.trim();
    if (!trimmed) {
      return {
        valid: false,
        error: 'Expression cannot be empty'
      };
    }

    // 4. Check allowed characters
    if (!this.ALLOWED_CHARS.test(trimmed)) {
      return {
        valid: false,
        error: 'Expression contains invalid characters. Only alphanumeric characters and mathematical operators are allowed.'
      };
    }

    // 5. Check for dangerous patterns
    for (const pattern of this.DANGEROUS_PATTERNS) {
      if (pattern.test(trimmed)) {
        return {
          valid: false,
          error: 'Expression contains disallowed patterns that could be unsafe'
        };
      }
    }

    // 6. Basic BC syntax validation (optional, BC will also catch this)
    const validation = this.validateBCSyntax(trimmed);
    if (!validation.valid) {
      return validation;
    }

    return {
      valid: true,
      sanitized: trimmed
    };
  }

  /**
   * Perform basic BC syntax validation
   * @param expression The expression to validate
   * @returns ValidationResult
   */
  private static validateBCSyntax(expression: string): ValidationResult {
    // Check for balanced parentheses
    let parenCount = 0;
    let braceCount = 0;
    let bracketCount = 0;

    for (const char of expression) {
      if (char === '(') parenCount++;
      if (char === ')') parenCount--;
      if (char === '{') braceCount++;
      if (char === '}') braceCount--;
      if (char === '[') bracketCount++;
      if (char === ']') bracketCount--;

      // Check for negative counts (closing before opening)
      if (parenCount < 0) {
        return {
          valid: false,
          error: 'Unbalanced parentheses: closing parenthesis before opening'
        };
      }
      if (braceCount < 0) {
        return {
          valid: false,
          error: 'Unbalanced braces: closing brace before opening'
        };
      }
      if (bracketCount < 0) {
        return {
          valid: false,
          error: 'Unbalanced brackets: closing bracket before opening'
        };
      }
    }

    // Check final counts
    if (parenCount !== 0) {
      return {
        valid: false,
        error: `Unbalanced parentheses: ${parenCount > 0 ? 'missing closing' : 'extra closing'} parenthesis`
      };
    }
    if (braceCount !== 0) {
      return {
        valid: false,
        error: `Unbalanced braces: ${braceCount > 0 ? 'missing closing' : 'extra closing'} brace`
      };
    }
    if (bracketCount !== 0) {
      return {
        valid: false,
        error: `Unbalanced brackets: ${bracketCount > 0 ? 'missing closing' : 'extra closing'} bracket`
      };
    }

    return { valid: true };
  }

  /**
   * Sanitize an expression (currently just trims whitespace)
   * Additional sanitization can be added if needed
   * @param expression The expression to sanitize
   * @returns Sanitized expression
   */
  static sanitize(expression: string): string {
    return expression.trim();
  }
}

File: src/python/bc_calculator_mcp/input_validator.py
"""
Input validation and sanitization module for BC Calculator
Prevents command injection and ensures safe input to BC
"""

import re
from typing import Pattern

from .types import ValidationResult


class InputValidator:
    """Validator for BC calculator expressions with security checks"""
    
    # Maximum expression length (10KB)
    MAX_LENGTH: int = 10000
    
    # Allowed characters: numbers, letters, math operators, BC syntax
    ALLOWED_CHARS: Pattern[str] = re.compile(r'^[0-9a-zA-Z+\-*/^().,;\s=<>!&|%{}\[\]]+$')
    
    # Dangerous patterns that could lead to command injection or system abuse
    DANGEROUS_PATTERNS: list[Pattern[str]] = [
        re.compile(r'system\s*\(', re.IGNORECASE),  # system() calls
        re.compile(r'exec\s*\(', re.IGNORECASE),    # exec() calls
        re.compile(r'`'),                            # Backticks (command substitution)
        re.compile(r'\$\('),                         # Command substitution
        re.compile(r'>\s*[/\w]'),                    # File redirects (output)
        re.compile(r'<\s*[/\w]'),                    # File redirects (input)
        re.compile(r'\|\s*[/\w]'),                   # Pipes to other commands
        re.compile(r';\s*(?:bash|sh|rm|cat|ls|pwd|chmod|chown)', re.IGNORECASE),  # Dangerous shell commands
    ]
    
    @classmethod
    def validate(cls, expression: str) -> ValidationResult:
        """
        Validate an expression for safety and correctness
        
        Args:
            expression: The expression to validate
            
        Returns:
            ValidationResult with validation status and optional error message
        """
        # 1. Check for null/undefined/empty
        if not expression:
            return ValidationResult(
                valid=False,
                error="Expression cannot be null or empty"
            )
        
        # 2. Check length
        if len(expression) > cls.MAX_LENGTH:
            return ValidationResult(
                valid=False,
                error=f"Expression too long (max {cls.MAX_LENGTH} characters, got {len(expression)})"
            )
        
        # 3. Check for empty input after trimming
        trimmed = expression.strip()
        if not trimmed:
            return ValidationResult(
                valid=False,
                error="Expression cannot be empty"
            )
        
        # 4. Check allowed characters
        if not cls.ALLOWED_CHARS.match(trimmed):
            return ValidationResult(
                valid=False,
                error="Expression contains invalid characters. Only alphanumeric characters and mathematical operators are allowed."
            )
        
        # 5. Check for dangerous patterns
        for pattern in cls.DANGEROUS_PATTERNS:
            if pattern.search(trimmed):
                return ValidationResult(
                    valid=False,
                    error="Expression contains disallowed patterns that could be unsafe"
                )
        
        # 6. Basic BC syntax validation
        validation = cls._validate_bc_syntax(trimmed)
        if not validation.valid:
            return validation
        
        return ValidationResult(
            valid=True,
            sanitized=trimmed
        )
    
    @classmethod
    def _validate_bc_syntax(cls, expression: str) -> ValidationResult:
        """
        Perform basic BC syntax validation
        
        Args:
            expression: The expression to validate
            
        Returns:
            ValidationResult
        """
        # Check for balanced parentheses, braces, and brackets
        paren_count = 0
        brace_count = 0
        bracket_count = 0
        
        for char in expression:
            if char == '(':
                paren_count += 1
            elif char == ')':
                paren_count -= 1
            elif char == '{':
                brace_count += 1
            elif char == '}':
                brace_count -= 1
            elif char == '[':
                bracket_count += 1
            elif char == ']':
                bracket_count -= 1
            
            # Check for negative counts (closing before opening)
            if paren_count < 0:
                return ValidationResult(
                    valid=False,
                    error="Unbalanced parentheses: closing parenthesis before opening"
                )
            if brace_count < 0:
                return ValidationResult(
                    valid=False,
                    error="Unbalanced braces: closing brace before opening"
                )
            if bracket_count < 0:
                return ValidationResult(
                    valid=False,
                    error="Unbalanced brackets: closing bracket before opening"
                )
        
        # Check final counts
        if paren_count != 0:
            return ValidationResult(
                valid=False,
                error=f"Unbalanced parentheses: {'missing closing' if paren_count > 0 else 'extra closing'} parenthesis"
            )
        if brace_count != 0:
            return ValidationResult(
                valid=False,
                error=f"Unbalanced braces: {'missing closing' if brace_count > 0 else 'extra closing'} brace"
            )
        if bracket_count != 0:
            return ValidationResult(
                valid=False,
                error=f"Unbalanced brackets: {'missing closing' if bracket_count > 0 else 'extra closing'} bracket"
            )
        
        return ValidationResult(valid=True)
    
    @classmethod
    def sanitize(cls, expression: str) -> str:
        """
        Sanitize an expression (currently just trims whitespace)
        Additional sanitization can be added if needed
        
        Args:
            expression: The expression to sanitize
            
        Returns:
            Sanitized expression
        """
        return expression.strip()